### PR TITLE
refactor(core): Throw an error when the document is not initialized.

### LINKED ISSUE #50138: Angular 16 migration: TypeError: Cannot read properties of undefined (reading 'getElementById')
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

No

### Description

Migrating from angular 15.2.9 to 16.0.0 we encounter the stack trace below when running the universal server (dev & prod build).
spa dev mode works as expected without any noticeable issues. 
Following the stack trace it's this piece of code that is encountering an issue:
```js
function retrieveTransferredState(doc,appId){const script=doc.getElementById(appId+"-state");
```
The error occurs whether we provide client hydration or not. 
We use TransferState throughout the application to send data from ssr to the client. 

EnvironmentService, that seems to be the root of this error, injects TransferState in the constructor.

I don't quite know how to drill down on this issue, since this seems to be happening very early in app lifecycle. Any Idea on how to create a reproduction would be highly appreciated. 

### Please provide a link to a minimal reproduction of the bug

_No response_

### Please provide the exception or error you saw

```true
TypeError: Cannot read properties of undefined (reading 'getElementById')
    at retrieveTransferredState (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2325653)
    at Object.initTransferState [as factory] (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2324438)
    at R3Injector.hydrate (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2320111)
    at R3Injector.get (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2318370)
    at injectInjectorOnly (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2186545)
    at Object.ɵɵinject (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2186715)
    at Object.EnvironmentService_Factory [as factory] (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:27288)
    at R3Injector.hydrate (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2320111)
    at R3Injector.get (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2318370)
    at injectInjectorOnly (/home/marco/FollowAlice/referral/frontend/dist/referral/server/main.js:1:2186545)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.0
Node: 18.16.0
Package Manager: npm 9.5.1
OS: linux x64

Angular: 16.0.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, material-luxon-adapter
... platform-browser, platform-browser-dynamic, platform-server
... router, service-worker, youtube-player

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.0
@angular-devkit/build-angular   16.0.0
@angular-devkit/core            16.0.0
@angular-devkit/schematics      16.0.0
@nguniversal/builders           16.0.0
@nguniversal/express-engine     16.0.0
@schematics/angular             16.0.0
typescript                      4.9.5
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeanMeche**
In case the document is accessed but not available we should throw ASAP an error to prevent non explicit errors.

See #50138, where this issue happens. 

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

### CODE REVIEW

File: packages/core/src/render3/interfaces/document.ts [hunk 1]
```diff
@@ -47,6 +49,11 @@ export function getDocument(): Document {
   } else if (typeof document !== 'undefined') {
     return document;
   }
+
+  if(typeof ngDevMode === 'undefined' || ngDevMode && !getDocument()) {
```
Comment (alan-agius4) [inline packages/core/src/render3/interfaces/document.ts:53]: I think this is case we should always throw irrespective if we are in dev mode or not.

Comment (JeanMeche) [reply packages/core/src/render3/interfaces/document.ts:53]: Yeah, definitely.

**Comment by JeanMeche** [comment]
> This also does not solve the true problem described in #50138

Yes, I understand it is not a true fix, just a mere helper to look in the right direction<

File: packages/core/src/render3/interfaces/document.ts [hunk 2]
```diff
@@ -47,6 +49,14 @@ export function getDocument(): Document {
   } else if (typeof document !== 'undefined') {
     return document;
   }
+
```
Comment (alan-agius4) [inline packages/core/src/render3/interfaces/document.ts:52]: The logic from this points onwards can be deleted in favor of the throw. If is redundant to call `getDocument`.

**Comment by alan-agius4** [comment]
> > This also does not solve the true problem described in #50138
> 
> Yes, I understand it is not a true fix, just a mere helper to look in the right direction.

I suggest that in this case, the commit message `fixes` is removed.

File: packages/core/src/render3/interfaces/document.ts [hunk 3]
```diff
@@ -47,10 +49,15 @@ export function getDocument(): Document {
...
+  throw new RuntimeError(
+      RuntimeErrorCode.MISSING_DOCUMENT,
+      (typeof ngDevMode === 'undefined' || ngDevMode) &&
+          `The document object is not available in this context. Make sure the DOCUMENT injection token is provided.`);
+
   // No "document" can be found. This should only happen if we are running ivy outside Angular and
```
Comment (alan-agius4) [inline packages/core/src/render3/interfaces/document.ts:58]: Nit: maybe move the comment before the throw.

**Review by alan-agius4 — APPROVED** [review]
LGTM

**Review by AndrewKushnir — APPROVED** [review]
@JeanMeche looks great, thanks! 👍 

Reviewed-for: fw-core, public-api

**Review by dylhunn — APPROVED** [review]
reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
[TGP](https://test.corp.google.com/OCL:529596253:BASE:529596204:1683258983931:752d192d).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: TGP is "green".

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 02147394880fe7b950d700f332758f5c09cb5c27.

**Pull request closed without merging**